File: cs/paths.py
"""Every well-known path in one place. All derived from HOME / env so tests can
point cs at a throwaway home."""
from __future__ import annotations

import os
from pathlib import Path


def home() -> Path:
    return Path(os.environ.get("HOME") or Path.home())


def claude_dir() -> Path:
    return Path(os.environ.get("CLAUDE_CONFIG_DIR") or home() / ".claude")


def claude_json() -> Path:
    """The monolithic ~/.claude.json (machine-local; cs only reads it)."""
    cfg = os.environ.get("CLAUDE_CONFIG_DIR")
    return (Path(cfg) if cfg else home()) / ".claude.json"


def cs_config_dir() -> Path:
    return Path(os.environ.get("CS_CONFIG_DIR") or home() / ".config" / "claude-share")


def machine_file() -> Path:
    return cs_config_dir() / "machine.toml"


def repo_dir() -> Path:
    """The private config repo clone (overridable in machine.toml via `repo`)."""
    return cs_config_dir() / "repo"


def state_dir() -> Path:
    base = Path(os.environ.get("XDG_STATE_HOME") or home() / ".local" / "state")
    return base / "cs"


def tool_root() -> Path:
    return Path(__file__).resolve().parent.parent


def templates_dir() -> Path:
    return tool_root() / "templates"


def expand(p: str) -> Path:
    """Expand ~ and env vars against the (possibly overridden) HOME."""
    p = os.path.expandvars(p)
    if p == "~" or p.startswith("~/"):
        p = str(home()) + p[1:]
    return Path(p)


def contract(p: Path) -> str:
    """Inverse of expand(): render a path with ~ for the home prefix."""
    try:
        rel = p.relative_to(home())
    except ValueError:
        return str(p)
    return "~" if str(rel) == "." else "~/" + str(rel)

File: src/paths.ts
import { homedir, hostname } from "node:os";
import { join, resolve, isAbsolute, relative } from "node:path";

export const home = () => process.env.HOME || homedir();
export const claudeDir = () => process.env.CLAUDE_CONFIG_DIR || join(home(), ".claude");
export const claudeJson = () => join(process.env.CLAUDE_CONFIG_DIR || home(), ".claude.json");
export const csConfigDir = () => process.env.CS_CONFIG_DIR || join(home(), ".config", "claude-share");
export const machineFile = () => join(csConfigDir(), "machine.toml");
export const shareDirDefault = () => join(csConfigDir(), "repo");
export const handoffStateDir = () => join(stateDir(), "handoff");
export const stateDir = () => join(process.env.XDG_STATE_HOME || join(home(), ".local", "state"), "cs");
export const toolRoot = () => resolve(new URL(".", import.meta.url).pathname, "..");
export const templatesDir = () => join(toolRoot(), "templates");
export const nodename = () => hostname();

/** ~ and $VAR expansion against the (possibly overridden) HOME. */
export function expand(p: string): string {
  let s = p.replace(/\$([A-Za-z_][A-Za-z0-9_]*)/g, (_, v) => process.env[v] ?? "");
  if (s === "~" || s.startsWith("~/")) s = home() + s.slice(1);
  return s;
}
/** Inverse: render with ~ for the home prefix. */
export function contract(p: string): string {
  const h = home();
  if (p === h) return "~";
  if (p.startsWith(h + "/")) return "~/" + p.slice(h.length + 1);
  return p;
}
export const isUnder = (child: string, parent: string) => { const r = relative(parent, child); return !!r && !r.startsWith("..") && !isAbsolute(r); };
